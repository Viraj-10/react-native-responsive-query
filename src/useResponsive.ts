import React from 'react';
import type { UseResponsiveProps, UseResponsiveReturnType } from './types';
import { StyleSheet, useWindowDimensions } from 'react-native';

export const useResponsive = (
  queries: UseResponsiveProps
): UseResponsiveReturnType => {
  const windowWidth = useWindowDimensions().width;

  const values = React.useMemo(() => {
    let styles = queries.initial
      ? [StyleSheet.create({ initial: queries.initial }).initial]
      : [];

    queries.query.forEach((queryRule) => {
      if (queryRule.style) {
        if (
          typeof queryRule.maxWidth === 'number' &&
          typeof queryRule.minWidth === 'number'
        ) {
          if (
            windowWidth >= queryRule.minWidth &&
            windowWidth <= queryRule.maxWidth
          ) {
            styles.push(
              StyleSheet.create({ rangeStyle: queryRule.style }).rangeStyle
            );
          }
        } else if (typeof queryRule.minWidth === 'number') {
          if (windowWidth >= queryRule.minWidth) {
            styles.push(
              StyleSheet.create({ minWidthStyle: queryRule.style })
                .minWidthStyle
            );
          }
        } else if (typeof queryRule.maxWidth === 'number') {
          if (windowWidth <= queryRule.maxWidth) {
            styles.push(
              StyleSheet.create({ maxWidthStyle: queryRule.style })
                .maxWidthStyle
            );
          }
        }
      }
    });

    return { styles };
  }, [queries, windowWidth]);

  return values;
};
