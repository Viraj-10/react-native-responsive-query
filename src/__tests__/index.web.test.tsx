import React from 'react';
import { useResponsive } from '../useResponsive.web';
import { render } from '@testing-library/react-native';
//@ts-ignore
import styleResolver from 'react-native-web/dist/exports/StyleSheet/styleResolver';
import { View } from 'react-native';

describe('test responsive styles', () => {
  it('verifies min width query', () => {
    const App = () => {
      const { styles } = useResponsive({
        initial: {
          backgroundColor: 'yellow',
        },
        query: [
          {
            minWidth: 500,
            style: {
              backgroundColor: 'black',
            },
          },
        ],
      });

      return <View testID="test" style={styles} />;
    };

    const { getByTestId } = render(<App />);
    const view = getByTestId('test');
    expect(view.props.style).toEqual([{ backgroundColor: 'yellow' }]);
    expect(styleResolver.sheet.getTextContent()).toContain(
      '/*[data-min-width-500-r-backgroundcolor-kemksi]{}*/@media only screen and (min-width: 500px) { [data-min-width-500-r-backgroundcolor-kemksi]{background-color:rgba(0,0,0,1.00);} }'
    );
  });

  it('verifies min width and max width query', () => {
    const App = () => {
      const { styles } = useResponsive({
        initial: {
          backgroundColor: 'yellow',
        },
        query: [
          {
            maxWidth: 400,
            style: {
              backgroundColor: 'black',
            },
          },
          {
            minWidth: 400,
            style: {
              backgroundColor: 'pink',
            },
          },
        ],
      });

      return <View testID="test" style={styles} />;
    };

    const { getByTestId } = render(<App />);
    const view = getByTestId('test');
    expect(view.props.style).toEqual([{ backgroundColor: 'yellow' }]);
    expect(styleResolver.sheet.getTextContent()).toContain(
      `/*[data-max-width-400-r-backgroundcolor-kemksi]{}*/@media only screen and (max-width: 400px) { [data-max-width-400-r-backgroundcolor-kemksi]{background-color:rgba(0,0,0,1.00);} }`
    );
    expect(styleResolver.sheet.getTextContent()).toContain(
      `/*[data-min-width-400-r-backgroundcolor-bwrkcz]{}*/@media only screen and (min-width: 400px) { [data-min-width-400-r-backgroundcolor-bwrkcz]{background-color:rgba(255,192,203,1.00);} }`
    );
  });
});
