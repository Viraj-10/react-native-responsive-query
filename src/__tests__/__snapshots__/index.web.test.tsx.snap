// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test responsive styles verifies min width and max width array queries 1`] = `
"[stylesheet-group=\\"0\\"]{}
html{-ms-text-size-adjust:100%;-webkit-text-size-adjust:100%;-webkit-tap-highlight-color:rgba(0,0,0,0);}
body{margin:0;}
button::-moz-focus-inner,input::-moz-focus-inner{border:0;padding:0;}
input::-webkit-search-cancel-button,input::-webkit-search-decoration,input::-webkit-search-results-button,input::-webkit-search-results-decoration{display:none;}
[stylesheet-group=\\"0.1\\"]{}
:focus:not([data-focusvisible-polyfill]){outline: none;}
[stylesheet-group=\\"3\\"]{}
/*d8jal210ecnfa{}*/@media only screen and (max-width: 400px) { [data-max-width-400$=\\"d8jal210ecnfa\\"]{background-color:rgba(0,0,0,1.00);}[data-max-width-400$=\\"d8jal210ecnfa\\"]{max-height:200px;} }
/*d8jal21i9q1lj{}*/@media only screen and (min-width: 400px) { [data-min-width-400$=\\"d8jal21i9q1lj\\"]{background-color:rgba(255,192,203,1.00);} }"
`;

exports[`test responsive styles verifies min width and max width query 1`] = `
"[stylesheet-group=\\"0\\"]{}
html{-ms-text-size-adjust:100%;-webkit-text-size-adjust:100%;-webkit-tap-highlight-color:rgba(0,0,0,0);}
body{margin:0;}
button::-moz-focus-inner,input::-moz-focus-inner{border:0;padding:0;}
input::-webkit-search-cancel-button,input::-webkit-search-decoration,input::-webkit-search-results-button,input::-webkit-search-results-decoration{display:none;}
[stylesheet-group=\\"0.1\\"]{}
:focus:not([data-focusvisible-polyfill]){outline: none;}
[stylesheet-group=\\"3\\"]{}
/*14jiznnp101yt{}*/@media only screen and (max-width: 400px) { [data-max-width-400$=\\"14jiznnp101yt\\"]{background-color:rgba(0,0,0,1.00);} }
/*14jiznn1i9q1lj{}*/@media only screen and (min-width: 400px) { [data-min-width-400$=\\"14jiznn1i9q1lj\\"]{background-color:rgba(255,192,203,1.00);} }"
`;

exports[`test responsive styles verifies min width query 1`] = `
"[stylesheet-group=\\"0\\"]{}
html{-ms-text-size-adjust:100%;-webkit-text-size-adjust:100%;-webkit-tap-highlight-color:rgba(0,0,0,0);}
body{margin:0;}
button::-moz-focus-inner,input::-moz-focus-inner{border:0;padding:0;}
input::-webkit-search-cancel-button,input::-webkit-search-decoration,input::-webkit-search-results-button,input::-webkit-search-results-decoration{display:none;}
[stylesheet-group=\\"0.1\\"]{}
:focus:not([data-focusvisible-polyfill]){outline: none;}
[stylesheet-group=\\"3\\"]{}
/*1s3271mu10ss8{}*/@media only screen and (min-width: 500px) { [data-min-width-500$=\\"1s3271mu10ss8\\"]{background-color:rgba(0,0,0,1.00);} }"
`;
